clients/stellar-relay/src/connection/services.rs

```diff
@@ -0,0 +1,296 @@
+
+				if expect_msg_len == 0 {
+					// there's nothing to read; wait for the next iteration
+					log::info!("WARNING!!! NOTHING TO READ!!!!");
+					continue
+				}
+
+				// let's start reading the actual stellar message.
+				readbuf = vec![0; expect_msg_len];
+
+				read_message(
+					&mut r_stream,
+					&tx_stream_reader,
+					&mut lack_bytes_from_prev,
+					&mut proc_id,
+					&mut readbuf,
+					expect_msg_len,
+				)
+				.await?;
+			},
+
+			Ok(Ok(_)) => {
+				retry = 0;
+				// let's read the continuation number of bytes from the previous message.
+				read_unfinished_message(
+					&mut r_stream,
+					&tx_stream_reader,
+					&mut lack_bytes_from_prev,
+					&mut proc_id,
+					&mut readbuf,
+				)
+				.await?;
+			},
+			Ok(Err(e)) => {
+				log::error!("read ERROR! {:?}", e);
+				return Err(Error::ConnectionFailed(e.to_string()))
+			},
+			Err(elapsed) => {
+				log::error!(
+					"{} for reading messages from Stellar Node. Retry: {}",
+					elapsed.to_string(),
+					retry
+				);
+
+				if retry >= retries {
+					return Err(Error::ConnectionFailed(format!(
+						"TIMED OUT reading for messages from the stream"
+					)))
+				}
+				retry += 1;
+			},
+		}
+	}
+}
+
+async fn _connection_handler(
+	actions: ConnectorActions,
+	conn: &mut Connector,
+	w_stream: &mut tcp::OwnedWriteHalf,
+) -> Result<(), Error> {
+	match actions {
+		// start the connection to Stellar node with a 'hello'
+		ConnectorActions::SendHello => {
+			log::info!("Starting Handshake with Hello.");
+			let msg = conn.create_hello_message(time_now())?;
+			w_stream.write_all(&msg).await.map_err(|e| Error::WriteFailed(e.to_string()))?;
+		},
+
+		// write message to the stream
+		ConnectorActions::SendMessage(msg) => {
+			let xdr_msg = conn.create_xdr_message(msg)?;
+			w_stream
+				.write_all(&xdr_msg)
+				.await
+				.map_err(|e| Error::WriteFailed(e.to_string()))?;
+		},
+
+		// handle incoming message from the stream
+		ConnectorActions::HandleMessage(xdr) => {
+			conn.process_raw_message(xdr).await?;
+		},
+	}
+
+	Ok(())
+}
+
+/// Handles actions for the connection.
+/// # Arguments
+/// * `conn` - the Connector that would send/handle messages to/from Stellar Node
+/// * `receiver` - The receiver for actions that the Connector should do.
+/// * `w_stream` -> the write half of the TcpStream to connect to the Stellar Node
+pub(crate) async fn connection_handler(
+	mut conn: Connector,
+	mut receiver: mpsc::Receiver<ConnectorActions>,
+	mut w_stream: tcp::OwnedWriteHalf,
+) -> Result<(), Error> {
+	let mut retry = 0;
+	loop {
+		match timeout(Duration::from_secs(conn.timeout_in_secs), receiver.recv()).await {
+			Ok(Some(action)) => {
+				retry = 0;
+				_connection_handler(action, &mut conn, &mut w_stream).await?;
+			},
+
+			Ok(None) => {},
+
+			Err(elapsed) => {
+				log::error!("{} for receiving messages. Retry: {}", elapsed.to_string(), retry);
+				if retry >= conn.retries {
```

Comment:
You are correct, this part here is just waiting for messages three times.
The logic for reconnection is in [overlay_connection.rs](https://github.com/pendulum-chain/spacewalk/pull/108/files#diff-5eeebd45117670e7260f1fd4cc2803213d9903a147c15c90f5720a2d752a33c9R51-R69) (better name!)
I will add comments on that part. :pray:

Reply:
@ebma  done adding comments. I also updated the README.md

Reply:
about the receiver, ah yes, it's can't be cloned. :pensive: MPSC.  There's the watch and/or broadcast channel.  :eyes:

Reply:
Thank you, I actually didn't think about changing the README 😅
I changed one more variable name and introduced a local variable for retries (because decreasing the `max_retries` field is counter-intuitive IMO). Can you check if it makes sense to you?